src/Features/Core/Portable/ForeachToFor/AbstractForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,346 @@
+                if (countSymbol == null || indexerSymbol == null)
+                {
+                    continue;
+                }
+
+                var countImpl = collectionType.FindImplementationForInterfaceMember(countSymbol) as IMethodSymbol;
+                var indexerImpl = collectionType.FindImplementationForInterfaceMember(indexerSymbol) as IMethodSymbol;
+                if (countImpl == null || indexerImpl == null)
+                {
+                    continue;
+                }
+
+                if (!semanticFact.IsAssignableTo(indexerImpl.ReturnType, foreachType, model.Compilation))
+                {
+                    continue;
+                }
+
+                // implicitly implemented!
+                if (countImpl.ExplicitInterfaceImplementations.IsEmpty &&
+                    indexerImpl.ExplicitInterfaceImplementations.IsEmpty)
+                {
+                    explicitCast = null;
+                    countName = Count;
+                    return;
+                }
+
+                if (explicitInterface == null)
+                {
+                    explicitInterface = current.ToDisplayString();
+                }
+            }
+
+            // okay, we don't have implicitly implemented one, but we do have explicitly implemented one
+            if (explicitInterface != null)
+            {
+                explicitCast = explicitInterface;
+                countName = Count;
+            }
+        }
+
+        private static bool IsNullOrErrorType(ITypeSymbol type)
+        {
+            return type == null || type is IErrorTypeSymbol;
+        }
+
+        private static IMethodSymbol GetInterfaceMembers(ITypeSymbol interfaceType, string memberName)
+        {
+            var members = interfaceType.GetMembers(memberName);
+            if (!members.IsEmpty)
+            {
+                return (IMethodSymbol)members[0];
+            }
+
+            foreach (var current in interfaceType.AllInterfaces)
+            {
+                var member = GetInterfaceMembers(current, memberName);
+                if (member != null)
+                {
+                    return member;
+                }
+            }
+
+            return null;
+        }
+
+        private static bool CheckRequireCollectionStatement(IOperation operation)
+        {
+            return operation.Kind != OperationKind.LocalReference && operation.Kind != OperationKind.FieldReference;
+        }
+
+        private IOperation RemoveImplicitConversion(IOperation collection)
+        {
+            if (collection is IConversionOperation conversion && conversion.IsImplicit)
+            {
+                return RemoveImplicitConversion(conversion.Operand);
+            }
+
+            return collection;
+        }
+
+        private bool CheckForeachVariable(SemanticModel semanticModel, ISymbol foreachVariable, SyntaxNode foreachStatement)
```

Comment:
this method name and return value are strange (esp given the code and comment inside).  Sounds like this should be called CheckIfForEachVariableIsNotWrittenInside.  Which is also weird since hte caller negates the value.    This should likely be CheckForeachVariableIsWrittenInside, and have the caller just use that directly without negation.

Reply:
sure.